pallets/stateful-storage/src/stateful_child_tree.rs

```diff
@@ -1,37 +1,54 @@
+use core::marker::PhantomData;
+
 use codec::{Decode, Encode};
 use common_primitives::msa::MessageSourceId;
 use frame_support::{
 	storage::{child, child::ChildInfo, ChildTriePrefixIterator},
-	Identity, StorageHasher,
+	Identity, ReversibleStorageHasher,
 };
-use sp_core::hexdisplay::AsBytesRef;
-use sp_std::prelude::*;
+use sp_std::{fmt::Debug, prelude::*};
 
-pub type StatefulPageKeyPart = Vec<u8>;
+pub trait IsStatefulKey: Debug + Encode + Decode + Eq + PartialEq {}
 
-/// Paginated Stateful data access utility
-pub struct StatefulChildTree;
-impl StatefulChildTree {
-	pub fn concat_keys(keys: &[StatefulPageKeyPart]) -> Vec<u8> {
-		let mut key = Vec::<u8>::new();
-		for k in keys {
-			let mut k2 = k.to_owned();
-			key.append(&mut k2);
-		}
-
-		key
-	}
+impl IsStatefulKey for () {}
+impl<T1: Debug + Decode + Encode + Eq + PartialEq> IsStatefulKey for (T1,) {}
+impl<
+		T1: Debug + Decode + Encode + Eq + PartialEq,
+		T2: Debug + Decode + Encode + Eq + PartialEq,
+	> IsStatefulKey for (T1, T2)
+{
+}
+impl<
+		T1: Debug + Decode + Encode + Eq + PartialEq,
+		T2: Debug + Decode + Encode + Eq + PartialEq,
+		T3: Debug + Decode + Encode + Eq + PartialEq,
+	> IsStatefulKey for (T1, T2, T3)
+{
+}
+impl<
+		T1: Debug + Decode + Encode + Eq + PartialEq,
+		T2: Debug + Decode + Encode + Eq + PartialEq,
+		T3: Debug + Decode + Encode + Eq + PartialEq,
+		T4: Debug + Decode + Encode + Eq + PartialEq,
+	> IsStatefulKey for (T1, T2, T3, T4)
+{
+}
 
+/// Paginated Stateful data access utility
+pub struct StatefulChildTree<H: ReversibleStorageHasher = Identity> {
+	hasher: PhantomData<H>,
+}
+impl<H: ReversibleStorageHasher> StatefulChildTree<H> {
 	/// Reads a child tree node and tries to decode it
 	///
 	/// The read is performed from the `msa_id` only. The `key` is not required. If the
 	/// data doesn't store under the given `key` `None` is returned.
-	pub fn try_read<V: Decode + Sized>(
+	pub fn try_read<K: IsStatefulKey, V: Decode + Sized>(
 		msa_id: &MessageSourceId,
-		keys: &[StatefulPageKeyPart],
+		keys: &K,
 	) -> Result<Option<V>, ()> {
 		let child = Self::get_child_tree(*msa_id);
-		let value = child::get_raw(&child, &Identity::hash(Self::concat_keys(keys).as_bytes_ref()));
+		let value = child::get_raw(&child, &H::hash(&keys.encode()).as_ref());
```

Comment:
I don't think this is going to work. We need to hash each key separately and concat those hashes. To be able to look them up by prefix

Reply:
Ah, right. Duh. `hash(x) + hash(y) <> hash(x + y)`